### PR TITLE
MAT-6401 missing context error

### PR DESCRIPTION
**Pull request by sb-cecilialiu**
## CQL to ELM Translation Service PR

Jira Ticket: [MAT-6401](https://jira.cms.gov/browse/MAT-6401)
(Optional) Related Tickets:

### All Submissions
* [x] This PR has the JIRA linked.
* [x] Required tests are included.
* [x] No extemporaneous files are included (i.e Complied files or testing results).
* [x] This PR is merging into the **correct branch**.
* [ ] All Documentation needed for this PR is Complete (or noted in a TODO or other Ticket).
* [ ] Any breaking changes or failing automations are noted by placing a comment on this PR.

### DevSecOps
If there is a question if this PR has a security or infrastructure impact, please contact the Security or DevOps engineer assigned to this project to discuss it further.

* [x] This PR has NO significant security impact (i.e Changing auth methods, Adding a new user type, Adding a required but vulnerable package).

### Reviewers
By Approving this PR you are attesting to the following:

*  Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose.
*  The tests appropriately test the new code, including edge cases.
*  If you have any concerns they are brought up either to the developer assigned, security engineer, or leads.

### CODE REVIEW

File: src/main/java/gov/cms/mat/cql_elm_translation/service/CqlConversionService.java [hunk 1]
```diff
@@ -180,4 +198,138 @@ private String formatMessage(CqlCompilerException e) {
...
+      cqlTranslator
+          .getExceptions()
+          .add(new MissingContextException(identifier, startLineForContext));
+    }
+  }
+
+  protected List<Integer> getStartLinesFromStatements(List<ExpressionDef> defs) {
+    List<Integer> startLines = new ArrayList<>();
+    if (CollectionUtils.isNotEmpty(defs)) {
+      for (ExpressionDef def : defs) {
```
Comment (adongare) [inline src/main/java/gov/cms/mat/cql_elm_translation/service/CqlConversionService.java]: Are we looping over all the definitions because the definitions are in random order? trying to understand why we need to iterate over all definitions.

File: src/main/java/gov/cms/mat/cql_elm_translation/service/CqlConversionService.java [hunk 2]
```diff
@@ -180,4 +198,138 @@ private String formatMessage(CqlCompilerException e) {
...
+        List<TrackBack> trackBacks = def.getTrackbacks();
+        if (CollectionUtils.isNotEmpty(trackBacks)) {
+          for (TrackBack trackBack : trackBacks) {
+            startLines.add(trackBack.getStartLine());
+          }
+        }
```
Comment (adongare) [inline src/main/java/gov/cms/mat/cql_elm_translation/service/CqlConversionService.java]: this can probably be its own private method that can be reused in all other line computing methods.

File: src/main/java/gov/cms/mat/cql_elm_translation/service/CqlConversionService.java [hunk 3]
```diff
@@ -180,4 +198,138 @@ private String formatMessage(CqlCompilerException e) {
...
+        List<TrackBack> trackBacks = def.getTrackbacks();
+        if (CollectionUtils.isNotEmpty(trackBacks)) {
+          for (TrackBack trackBack : trackBacks) {
+            endLines.add(trackBack.getEndLine());
+          }
+        }
+      }
+    }
+    return endLines;
+  }
```
Comment (adongare) [inline src/main/java/gov/cms/mat/cql_elm_translation/service/CqlConversionService.java:304]: the closer look at this gives some idea about common pattern, I think this can be simplified with just few lines and reused for all line methods

```
protected List<Integer> getLines(List<? extends Element> defs) {
    List<Integer> startLines = new ArrayList<>();
    if (CollectionUtils.isNotEmpty(defs)) {
      for (Element def : defs) {
        List<TrackBack> trackBacks = def.getTrackbacks();
        if (CollectionUtils.isNotEmpty(trackBacks)) {
          for (TrackBack trackBack : trackBacks) {
            startLines.add(trackBack.getStartLine());
          }
        }
      }
    }
    return startLines;
  }
 
```

File: src/main/java/gov/cms/mat/cql_elm_translation/service/CqlConversionService.java (hunk 1 above)
Comment (sb-cecilialiu) [reply src/main/java/gov/cms/mat/cql_elm_translation/service/CqlConversionService.java]: @adongare It seems to me that "context Patient" needs to be above definitions otherwise there will be error?

File: src/main/java/gov/cms/mat/cql_elm_translation/service/CqlConversionService.java (hunk 2 above)
Comment (sb-cecilialiu) [reply src/main/java/gov/cms/mat/cql_elm_translation/service/CqlConversionService.java]: @adongare refactored. Thanks!

File: src/main/java/gov/cms/mat/cql_elm_translation/service/CqlConversionService.java (hunk 3 above)
Comment (sb-cecilialiu) [reply src/main/java/gov/cms/mat/cql_elm_translation/service/CqlConversionService.java:304]: @adongare done, thanks~

**Review by adongare — APPROVED** [review]
